src/Features/CSharp/Portable/RemoveUnreachableCode/CSharpRemoveUnreachableCodeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnreachableCode
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpRemoveUnreachableCodeDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        private const string CS0162 = nameof(CS0162); // Unreachable code detected
+
+        public const string IsCascadedSection = nameof(IsCascadedSection);
+        private static readonly ImmutableDictionary<string, string> s_additionalProperties = ImmutableDictionary<string, string>.Empty.Add(IsCascadedSection, "");
+
+        public CSharpRemoveUnreachableCodeDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveUnreachableCodeDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Unreachable_code_detected), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeBlock, ImmutableArray.Create(SyntaxKind.Block));
+
+        private void AnalyzeBlock(SyntaxNodeAnalysisContext context)
+        {
+            var block = context.Node;
+            if (!IsOutermostBlock(block))
```

Comment:
So i'm still skeptical that this would be beneficial or actually account for much at all.  We already have analyzers that are going to run on *far* more types of nodes and actually have to do far more work.  To the point that this seems like this would be noise in comparison.  For example, the "string.Format" analyzer is going to have to run on *every* invocation (which dwarfs the number of blocks).

--

Furthermore, i'm still unconvinced that actually enumerating all the cases will be helpful.  That's because there are a *lot* of cases, and many of which we would still end up having to do checks to see if it's worth analyzing.  For example, we don't need to analyze anything like a method if it does *not* contain a block within it.  But if we analyze methoddeclarations we can't just see if it has a body or not.  That's because it might be an expression body that then contains a lambda somewhere within it that then has unreachable code inside that lambda.  

This will come up with fields, for example.  We'll have to compile all fields in case they might have an expressoin that contains a lambda in it.  In the current approach, we only bother looking if you actually have a block that could then have unreachable code in it.

--

Note: we could try to deal with this by also having lambdas be something we get called back on.  But then we'd need this same code to ensure htat we're not processing a lambda that we would have already processed because it was part of something larger.

--

Right now the code is both clear, and seems to require checks that should be negligible (esp. compared to what the analysis driver has to already deal with).   The added checks would actually be much more complex to get right, and it seems like it could definitely lead to more expensive work (i.e. having to check in the reverse direction, or having to compile things that are unnecessary).

I started going down this path but it's a *lot* of complexity for questionable gain IMO.